File: android/app/src/main/kotlin/com/edgelink/core/RelayAuth.kt
package com.edgelink.core

import kotlinx.serialization.Serializable
import java.util.Base64

@Serializable
data class RelayAuthEnvelope(
    val t: String = "relay.auth",
    val b: Body
) {
    @Serializable
    data class Body(
        val hostId: String,
        val deviceId: String,
        val ts: Long,
        val sig: String
    )
}

object RelayAuth {
    fun message(deviceId: String, timestampSeconds: Long): ByteArray =
        "EdgeLink relay auth v1\n$deviceId\n$timestampSeconds".encodeToByteArray()

    fun envelope(hostId: String, identity: LocalIdentity, timestampSeconds: Long, signature: ByteArray): RelayAuthEnvelope =
        RelayAuthEnvelope(
            b = RelayAuthEnvelope.Body(
                hostId = hostId,
                deviceId = identity.deviceId,
                ts = timestampSeconds,
                sig = Base64.getEncoder().encodeToString(signature)
            )
        )
}

File: mac/Sources/EdgeLinkKit/RelayAuth.swift
import CryptoKit
import Foundation

public struct RelayAuthEnvelope: Codable, Equatable, Sendable {
    public struct Body: Codable, Equatable, Sendable {
        public let hostId: String
        public let deviceId: String
        public let ts: Int64
        public let sig: String
    }

    public let t: String
    public let b: Body
}

public enum RelayAuth {
    public static func message(deviceId: String, timestampSeconds: Int64) -> Data {
        Data("EdgeLink relay auth v1\n\(deviceId)\n\(timestampSeconds)".utf8)
    }

    public static func envelope(hostId: String, identity: LocalIdentity, timestampSeconds: Int64) throws -> RelayAuthEnvelope {
        let signature = try identity.signingKey.signature(for: message(deviceId: identity.deviceId, timestampSeconds: timestampSeconds))
        return RelayAuthEnvelope(
            t: "relay.auth",
            b: .init(
                hostId: hostId,
                deviceId: identity.deviceId,
                ts: timestampSeconds,
                sig: signature.base64EncodedString()
            )
        )
    }
}
